Add timeoutTimer to dsmr integration

mjkl-gh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I've added a `timeoutTimer` to `connect_and_reconnect()` function to reset the connection whenever it hangs (most likely due to ungraceful disconnect)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40595 
- This PR is related to issue: #40595
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mjkl-gh committed:
Added timeoutTimer to dsmr integration

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

mjkl-gh committed:
Fix bugs from porting

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

mjkl-gh committed:
fix bug where updateTimer was restarted after disconnect

calling update_entities at the end to reflect unavailable entitities restarted the timeout timer

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

mjkl-gh committed:
change test from boolean to future

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

mjkl-gh committed:
Merge branch 'dev' of https://github.com/home-assistant/core into dev

mjkl-gh committed:
Fix linting

mjkl-gh committed:
remove old class file

mjkl-gh committed:
Merge branch 'dev' of https://github.com/home-assistant/core into dev

mjkl-gh committed:
Add tests for timeout_timer

mjkl-gh committed:
added Timer tests

mjkl-gh committed:
remove old test

mjkl-gh committed:
fix codespell

mjkl-gh commented:
@RobBie1221 I'm sorry to have to ask, But I'm stuck on the test specified at [line 542 of test_sensor.py](https://github.com/mjkl-gh/core/blob/580183962b7d6e132cd5ab3b7a8f2ed1b4044ebb/tests/components/dsmr/test_sensor.py#L541)

Even if I reset all my changes except changing 'await transport.wait_closed()'  to 
```
                    await asyncio.wait(
                        [protocol.wait_closed()],
                        return_when=asyncio.FIRST_COMPLETED,
                    )
```
It will fail with connection_factory.call_count = 1

I know from running this in my home this code functions and triggers a reconnect. (when using this with a tcp client, I have no serial device available to me at this time). So I'm stuck trying to figure out why this test fails. Do you maybe have any insight?

mjkl-gh commented:
Just figured out there is some sort of race condition happening. Adding `asyncio.sleep(1)` before ` assert connection_factory.call_count >= 2, "connecting not retried"` makes the test resolve correctly

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @invalid-email-address, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43825).

Once you do that we will be able to review and accept this pull request.

Thanks!

mjkl-gh committed:
Merge branch 'dev' of https://github.com/home-assistant/core into dev

mjkl-gh committed:
fix racecondition

mjkl-gh committed:
added test for tcp timeout

mjkl-gh committed:
add tests for timer class

mjkl-gh committed:
remove unnecessary code to reduce test requirement

mjkl-gh committed:
Merge branch 'dev' of https://github.com/home-assistant/core into fix-dsmr

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
             return f"{unit}/{TIME_HOURS}"
+
+
+class Timer:
```
Can you move this to `homeassistant/util/async_.py`

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+        self.timerhandle = None
+        self.set_timeout(delay)
+        self.set_callback(timer_callback)
+        self.args = args
+        if loop:
+            self.loop = loop
+        else:
+            self.loop = asyncio.get_event_loop()
```
Let's drop support for passing in `args` and `loop`. Use `asyncio.get_running_loop()` to get the loop.

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+        """
+        self.delay = delay
+
+    def set_callback(self, timer_callback):
```
I would just drop this method. Limit to functionality that is used.

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+
+    def start(self):
+        """Start the timer by creating a task to be called on."""
+        if not self._timer_active():
```
Use guard clause and don't raise `Exception` but instead raise more appropriate errors like `ValueError` or `RuntimeError`.

```python
if self._timer_active():
    raise RuntimeError("Timer already started")
```

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+        """Stop the running timer by canceling the task."""
+        if not self.timerhandle:
+            raise Exception("Timer was stopped before being started")
+        self.timerhandle.cancel()
```
```suggestion
        self.timerhandle.cancel()
        self.timerhandle = None
```

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+            and not self._timer_done()
+        )
+
+    def __del__(self):
```
I think that this is unexpected behavior. Better to require explicitly calling `stop`.

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -409,3 +444,105 @@ def unit_of_measurement(self):
+    def _timer_active(self):
+        return (
+            self.timerhandle
+            and not self.timerhandle.cancelled()
+            and not self._timer_done()
```
If you make sure you always clear the timer handle when it's done or cancelled, you can just check against `None` here.

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -150,6 +149,26 @@ async def async_setup_entry(
+    def timer_callback():
+        nonlocal update_timeout
+        nonlocal update_count
+        update_count += 1
```
Count is not used.

balloob commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -150,6 +149,26 @@ async def async_setup_entry(
+        if not update_timeout.done():
+            update_timeout.set_result(True)
+        else:
+            raise Exception("Future " "update_timeout" " Already set")
```
```suggestion
            raise RuntimeError("Future update_timeout already set")
```

balloob commented:
Ah yes, this is a lot better fixed upstream. Thanks for the ping @RobBie1221

mjkl-gh commented:
> @mjkl-gh Is this PR still relevant given that there is an upstream PR as well to solve the same thing? (https://github.com/ndokter/dsmr_parser/pull/73/files)
@RobBie1221 
No, I was waiting on the PR being merged before closing this. There are currently multiple PR's open on the upstream repo without maintainer input. I'm therefore slightly skeptical about it being merged, I'm afraid.
